src/Dofs/ConstraintHandler.jl

```diff
@@ -1587,3 +1587,186 @@ function __check_periodic_faces_f(grid::Grid, fi::FaceIndex, fj::FaceIndex, xmi,
+## Local application of constraints ##
+######################################
+
+"""
+    apply_local!(
+        local_matrix::AbstractMatrix, local_vector::AbstractVector,
+        global_dofs::AbstractVector, ch::ConstraintHandler;
+        apply_zero::Bool = false
+    )
+
+Similar to [`apply!`](@ref) but perform condensation of constrained degrees-of-freedom
+locally in `local_matrix` and `local_vector` *before* they are to be assembled into the
+global system.
+
+When the keyword argument `apply_zero` is `true` all inhomogeneities are set to `0` (cf.
+[`apply!`](@ref) vs [`apply_zero!`](@ref)).
+
+This method can only be used if all constraints are "local", i.e. no constraint couples with
+dofs outside of the element dofs (`global_dofs`) since condensation of such constraints
+requires writing to entries in the global matrix/vector. For such a case,
+[`apply_assemble!`](@ref) can be used instead.
+
+Note that this method is destructive since it, by definition, modifies `local_matrix` and
+`local_vector`.
+"""
+function apply_local!(local_matrix::AbstractMatrix, local_vector::AbstractVector,
+                      global_dofs::AbstractVector, ch::ConstraintHandler;
+                      apply_zero::Bool = false)
+    return _apply_local!(local_matrix, local_vector, global_dofs, ch, apply_zero,
+                         #=global_matrix=# nothing, #=global_vector=# nothing)
+end
+
+# Element local application of boundary conditions. Global matrix and vectors are necessary
+# if there are affine constraints that connect dofs from different elements.
+function _apply_local!(local_matrix::AbstractMatrix, local_vector::AbstractVector,
+                       global_dofs::AbstractVector, ch::ConstraintHandler, apply_zero::Bool,
+                       global_matrix, global_vector)
+    # TODO: With apply_zero it shouldn't be required to pass the vector.
+    length(global_dofs) == size(local_matrix, 1) == size(local_matrix, 2) == length(local_vector) || error("?")
+    # First pass over the dofs check whether there are any constrained dofs at all
+    has_constraints = false
+    has_nontrivial_affine_constraints = false
+    # 1. Adjust local vector
+    @inbounds for (local_dof, global_dof) in pairs(global_dofs)
+        # Check if this dof is constrained
+        pdofs_index = get(ch.dofmapping, global_dof, nothing)
+        pdofs_index === nothing && continue # Not constrained, move on
+        has_constraints = true
+        # Add inhomogeneities to local_vector: local_vector - local_matrix * inhomogeneities
+        v = ch.inhomogeneities[pdofs_index]
+        if !apply_zero && v != 0
+            for j in axes(local_matrix, 1)
+                local_vector[j] -= v * local_matrix[j, local_dof]
+            end
+        end
+        # Check if this is an affine constraint
+        if has_nontrivial_affine_constraints || (
+           coeffs = ch.dofcoefficients[pdofs_index];
+           !(coeffs === nothing || isempty(coeffs))
```

Comment:
Difficult to read when everything is in the if statement? :P 
Also, maybe we should check that coeffs is not empty in the close!(ch), so we don't have to check for that here?